src/components/editor/EditorLogin.jsx

```diff
@@ -1,74 +1,92 @@
-    });
-    this.handlePasswordChange = this.handlePasswordChange.bind(this);
-    this.handleSubmit = this.handleSubmit.bind(this);
+    this.state = {
+      gapiReady: false,
+    };
+    this.handleRedirect = this.handleRedirect.bind(this);
+    this.onSignIn = this.onSignIn.bind(this);
+    this.onSignInSuccess = this.onSignInSuccess.bind(this);
   }
 
-  handlePasswordChange(e) {
-    this.safeSetState({
-      passwordValue: e.target.value,
-    });
-  }
+  componentDidMount() {
+    const script = document.createElement('script');
+    script.src = 'https://apis.google.com/js/platform.js';
+    document.body.appendChild(script);
 
-  handleSubmit(e) {
-    e.preventDefault();
-    const pass = e.target.password.value;
-    // eslint-disable-next-line max-len
-    if (hash(pass) !== 'eaafc81d7868e1c203ecc90f387acfa4c24d1027134b0bfda6fd7c536efc5d8dd5718609a407dbfcd41e747aec331153d47733153afb7c125c558acba3fb6bcd') {
-      window.alert('Incorrect password');
-      e.target.password.value = ''; // eslint-disable-line no-param-reassign
-    } else {
-      // Otherwise it is correct and we let the site redirect the client
-      let url = this.props.location.query.url || '';
-      if (url) {
-        _.forEach(this.props.location.query, (q, name) => {
-          if (name !== 'url') {
-            url = `${url}&${name}=${q}`;
-          }
+    script.onload = () => {
+      window.gapi.load('auth2', () => {
+        this.setState({
+          gapiReady: true,
         });
+        if (!window.gapi.auth2.getAuthInstance()) {
+          window.gapi.auth2.init({
+            client_id: googleClientID,
+          });
+        }
+      });
+    };
+  }
+
+  handleRedirect() {
+    // Add additional query params to new url
+    let url = this.props.location.query.url || '';
+    if (url) {
+      _.forEach(this.props.location.query, (q, name) => {
+        if (name !== 'url') {
+          url = `${url}&${name}=${q}`;
+        }
+      });
+    }
+    browserHistory.push(url);
+  }
+
+  onSignInSuccess(response) {
+    const token = { data: response.getAuthResponse().id_token };
+
+    const xhr = new XMLHttpRequest();
+    xhr.open('POST', '/googlelogin');
+    xhr.setRequestHeader('Content-Type', 'application/json');
+
+    xhr.onreadystatechange = () => {
+      if (xhr.readyState === XMLHttpRequest.DONE && xhr.status === 200) {
+        this.handleRedirect();
+      } else if (xhr.readyState === XMLHttpRequest.DONE) {
+        alert(`Google Login Failed.\n${xhr.status}: ${xhr.statusText}`);
       }
-      browserHistory.push(url);
+    };
+
+    xhr.send(JSON.stringify(token));
+  }
+
+  onSignInFailure(response) {
+    alert(`Google Login Failed.\n${response.error}`);
+  }
+
+  onSignIn() {
+    if (this.state.gapiReady) {
+      const auth = window.gapi.auth2.getAuthInstance();
+      auth.signIn().then(res => this.onSignInSuccess(res), err => this.onSignInFailure(err));
+    } else {
+      alert('Google API is not ready. Try again.');
```

Comment:
Do you think we can disable the button until it is ready instead of this? I see that it's simply a div and maybe it's a bit convoluted how Google is implementing it, but even if you can't just put in a `disable` flag to the div one simple way should be to do something like
```js
let additionalStyles = { ...style.button };
if (!this.state.gapiReady) {
  additionalStyles = { ...additionalStyles, color: 'grey' };
}
return <div ... style={additionalStyles} />;
```

And then just plain remove the else here so that nothing happens when it's not ready, and the UX indicates it with the color

Reply:
sounds cool! thanks

Reply:
But yeah, it might be that Google already implemented the `disable` attribute, so definitely try that out as that's much cleaner of course. And also, just a small semi-correction to my code above, it's not necessary to do a full copy in the if statement like I did as we already did a copy, so you can just do `additionalStyles.color = 'grey';` if `disable` doesn't work.

Reply:
From what I've gathered Google does not provide `disable`. To make it worse, the `div` Google provides has multiple inner divs - setting the div itself's color doesn't change how the button looks (if we really wanna do it we'll have to manipulate the inner stylings somehow).  
On the other hand, from my experience using gapi, the delay is minimal and I don't notice it at all, so it may be trivial to set an alternative styling while it's loading for a fraction of a second. But of course the loading time will vary depending on the Internet connection, etc. 
I'm not sure, what do you think?

Reply:
Just tried in Chrome Dev Tools and something like `additionalStyles.opacity = '0.5'` works fine, and if you want to change something else like the background color you change this class: 
![screenshot from 2017-12-06 14-14-17](https://user-images.githubusercontent.com/18149242/33680529-e83fde32-da8f-11e7-8b81-4dd40663efaf.png)

By changing the editor tools stylesheet to include something like
```css
.g-signin2.google-login-disabled .abcRioButtonLightBlue {
  background-color: #newColor;
  color: #newTextColor;
}
```
For example.

And yeah probably it'll usually connect, but if you're gonna add the case in the code we should add it properly, and for good form it is probably also a good idea to add it.

Reply:
gotcha. I'll try your suggestions!